### PR TITLE
Fix the issue Support OpenMP critical #92

### PR DESCRIPTION
**Pull request by jedishao23**
Treat the openMP critical as a lock.

### CODE REVIEW

**Comment by BradSwain** [comment]
Hi @jedishao23 thanks for the PR. Looks very high quality!
I will try to do a full review later today and get this merged.

**Comment by BradSwain** [comment]
closes #92

File: src/IR/IRImpls.h [hunk 1]
```diff
@@ -207,6 +207,22 @@ class LockIRImpl : public LockIR {
   static inline bool classof(const IR *e) { return e->type == T; }
 };
 
+class OpenMPCriticalStart: public LockIR {
+  const unsigned int criticalEntryOffset = 2;
```
Comment (BradSwain) [inline src/IR/IRImpls.h:211]: Can you add a comment here describing the kmpc_critical API and linking to the functio in LLVM's github project?

You can use the comment for OpenMP fork as a template

https://github.com/coderrect-inc/OpenRace/blob/328fe090a12055e4e2731557b1ea1cc32ede8776/src/IR/IRImpls.h#L122-L129

File: src/IR/IRImpls.h [hunk 2]
```diff
@@ -236,6 +252,22 @@ class UnlockIRImpl : public UnlockIR {
   static inline bool classof(const IR *e) { return e->type == T; }
 };
 
+class OpenMPCriticalEnd : public UnlockIR {
+  const unsigned int criticalEndOffset = 2;
```
Comment (BradSwain) [inline src/IR/IRImpls.h:256]: Can you add a comment here describing the kmpc_end_critical API and linking to the function in LLVM's github project?

You can use the comment for OpenMP fork as a template.

https://github.com/coderrect-inc/OpenRace/blob/328fe090a12055e4e2731557b1ea1cc32ede8776/src/IR/IRImpls.h#L122-L129

**Review by BradSwain — CHANGES_REQUESTED** [review]
Look great! Just two small changes adding some comments about the OpenMP API's modeled and I think it is ready to be merged.

File: src/IR/IRImpls.h (hunk 1 above)
Comment (funemy) [reply src/IR/IRImpls.h:211]: I suggest we should rename `criticalEntryOffset`.

The offset seems to denote the actual lock object being acquired or just a identify for the critical section?
(according to the reference https://openmp.llvm.org/Reference.pdf)

In that case, we should name it `identifyOffset` or `lockOffset`.

File: src/IR/IRImpls.h (hunk 2 above)
Comment (funemy) [reply src/IR/IRImpls.h:256]: Same as my comment above. We should rename `criticalEndOffset`

Comment (jedishao23) [reply src/IR/IRImpls.h:256]: OK, I'll rename them and add comments.

**Review by funemy — COMMENTED** [review]
LGTM now, I'll let @BradSwain approve it

**Review by BradSwain — APPROVED** [review]

**Comment by BradSwain** [comment]
Thanks @jedishao23 looks good to me. Thanks again for the high quality contribution!

**Pull request merged**